### PR TITLE
1066 1067 1068: Number of Schools Supported by Development Partners

### PR DESCRIPTION
**Pull request by chris-bes**
### Issues: 
https://github.com/beyondessential/tupaia-backlog/issues/1066
https://github.com/beyondessential/tupaia-backlog/issues/1067
https://github.com/beyondessential/tupaia-backlog/issues/1068

### Changes:

- Added countCalculatedValuePerOrgUnit dataBuilder
- Implemented dev partners pie chart in laos schools

### CODE REVIEW

File: packages/database/src/migrations/20200930141002-AddLaosSchoolsNumberOfSchoolsSupportedByDevelopmentPartnersDashboardReport-modifies-data.js [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    name: 'Number of Secondary Schools Supported by Development Partners',
+    schoolType: 'Secondary',
+  },
+];
+
+const BASE_DATA_BUILDER_CONFIG = {
+  operation: {
+    operator: 'COMBINE_BINARY_AS_STRING',
+    dataElementToString: {
+      SchCVD022: {
```
Comment (Alastair-L) [inline packages/database/src/migrations/20200930141002-AddLaosSchoolsNumberOfSchoolsSupportedByDevelopmentPartnersDashboardReport-modifies-data.js:47]: I note that you have asked in the card about this, but personally I would expect that 'No support' would be displayed simply if none of the other codes are answered 'Yes'. Let's see what Megan gets back to us with.

File: packages/database/src/migrations/20200930141002-AddLaosSchoolsNumberOfSchoolsSupportedByDevelopmentPartnersDashboardReport-modifies-data.js [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+      operator: '=',
+      value: 'Other',
+    },
+    'No support': {
+      operator: 'regex',
+      value: 'No support',
+    },
+    Multiple: {
+      operator: 'regex',
+      value: '^(?=.*,)(?!.*No support).*', // multiple indicators except No support
```
Comment (Alastair-L) [inline packages/database/src/migrations/20200930141002-AddLaosSchoolsNumberOfSchoolsSupportedByDevelopmentPartnersDashboardReport-modifies-data.js:120]: Clever regex!

File: packages/database/src/migrations/20200930141002-AddLaosSchoolsNumberOfSchoolsSupportedByDevelopmentPartnersDashboardReport-modifies-data.js (hunk 1 above)
Comment (chris-bes) [reply packages/database/src/migrations/20200930141002-AddLaosSchoolsNumberOfSchoolsSupportedByDevelopmentPartnersDashboardReport-modifies-data.js:47]: @Alastair-L ah yes, Megan did get back to me on Slack end of last Friday. She's confirmed that we should use `SchCVD022` for No support

File: packages/web-config-server/src/apiV1/dataBuilders/helpers/calculateOperationForAnalytics.js [hunk 3]
```diff
@@ -92,6 +101,13 @@ const SINGLE_ANALYTIC_OPERATORS = ['CHECK_CONDITION', 'FORMAT', 'GROUP'];
 
 const ARITHMETIC_OPERATORS = ['DIVIDE', 'SUBTRACT'];
 
+export const getDataElementsFromCalculateOperationConfig = config =>
```
Comment (Alastair-L) [inline packages/web-config-server/src/apiV1/dataBuilders/helpers/calculateOperationForAnalytics.js:104]: 🎉

File: packages/web-config-server/src/apiV1/dataBuilders/helpers/calculateOperationForAnalytics.js [hunk 4]
```diff
@@ -92,6 +101,13 @@ const SINGLE_ANALYTIC_OPERATORS = ['CHECK_CONDITION', 'FORMAT', 'GROUP'];
...
+  typeof config === 'string'
+    ? config
```
Comment (Alastair-L) [inline packages/web-config-server/src/apiV1/dataBuilders/helpers/calculateOperationForAnalytics.js:106]: Currently, `config` can't actually be a string when it is passed into `calculateOperationForAnalytics`. That case is handled earlier, in `getCalculatedValuesByCell`. Perhaps the `config === 'string'` check could be done in the matrix builder in order to make more sense?

File: packages/web-config-server/src/apiV1/dataBuilders/generic/count/countCalculatedValuesPerOrgUnit.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    const dataElementCodes = getDataElementsFromCalculateOperationConfig(operation);
+    const { results } = await this.fetchAnalytics(dataElementCodes);
+    const analyticsByOrgUnit = groupBy(results, 'organisationUnit');
+    const calculatedValues = Object.values(analyticsByOrgUnit).map(analytics =>
+      calculateOperationForAnalytics(analytics, operation),
+    );
+
+    const data = Object.entries(dataClasses).map(([dataClass, condition]) => {
+      let count = 0;
+      calculatedValues.forEach(value => {
```
Comment (Alastair-L) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/count/countCalculatedValuesPerOrgUnit.js:21]: Could be a tiny bit shorter with `.filter` `.length`, but up to you if you think it is more or less readable.

File: packages/web-config-server/src/apiV1/dataBuilders/generic/count/countCalculatedValuesPerOrgUnit.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        value: count,
+      };
+    });
+
+    return { data };
+  }
+}
+
+export const countCalculatedValuesPerOrgUnit = async (
+  { dataBuilderConfig, query, entity },
```
Comment (Alastair-L) [inline packages/web-config-server/src/apiV1/dataBuilders/generic/count/countCalculatedValuesPerOrgUnit.js:38]: @edmofro, Is this a model conflict?

**Review by Alastair-L — CHANGES_REQUESTED** [review]
Nice work! Just a few small comments to respond to 😄

File: packages/database/src/migrations/20200930141002-AddLaosSchoolsNumberOfSchoolsSupportedByDevelopmentPartnersDashboardReport-modifies-data.js (hunk 1 above)
Comment (Alastair-L) [reply packages/database/src/migrations/20200930141002-AddLaosSchoolsNumberOfSchoolsSupportedByDevelopmentPartnersDashboardReport-modifies-data.js:47]: Cool. If Megan's happy with it so am I 😄

File: packages/web-config-server/src/apiV1/dataBuilders/generic/count/countCalculatedValuesPerOrgUnit.js (hunk 6 above)
Comment (edmofro) [reply packages/web-config-server/src/apiV1/dataBuilders/generic/count/countCalculatedValuesPerOrgUnit.js:38]: Yep, thanks @Alastair-L! Added to my list.

File: packages/web-config-server/src/apiV1/dataBuilders/helpers/calculateOperationForAnalytics.js (hunk 4 above)
Comment (chris-bes) [reply packages/web-config-server/src/apiV1/dataBuilders/helpers/calculateOperationForAnalytics.js]: I totally agree!

**Review by Alastair-L — APPROVED** [review]
LGTM!

**Comment by Alastair-L** [comment]
@chris-bes, could you please check out the last few commits?

**Comment by chris-bes** [comment]
@Alastair-L Looks good! (I can't approve it cause I created this PR)

**Comment by Alastair-L** [comment]
Thanks Chris :)

File: packages/web-config-server/src/apiV1/dataBuilders/generic/count/countCalculatedValuesPerOrgUnit.js (hunk 6 above)
Comment (edmofro) [reply packages/web-config-server/src/apiV1/dataBuilders/generic/count/countCalculatedValuesPerOrgUnit.js:43]: Resolved in https://github.com/beyondessential/tupaia/pull/1484/commits/128b6edfc558a68c621c983061d0f46b300a19c2

**Pull request merged**